v-next/core/test/internal/hook-manager.ts

```diff
@@ -0,0 +1,938 @@
+import type {
+  ConfigurationVariable,
+  HardhatConfig,
+  HardhatUserConfig,
+} from "../../src/types/config.js";
+import type {
+  ConfigHooks,
+  HardhatRuntimeEnvironmentHooks as HreHooks,
+  HardhatUserConfigValidationError,
+  HookContext,
+  HookManager,
+} from "../../src/types/hooks.js";
+import type { HardhatRuntimeEnvironment } from "../../src/types/hre.js";
+import type { HardhatPlugin } from "../../src/types/plugins.js";
+import type { Task, TaskManager } from "../../src/types/tasks.js";
+import type { UserInterruptionManager } from "../../src/types/user-interruptions.js";
+
+import assert from "node:assert/strict";
+import { describe, it, beforeEach } from "node:test";
+
+import { HookManagerImplementation } from "../../src/internal/hook-manager.js";
+import { UserInterruptionManagerImplementation } from "../../src/internal/user-interruptions.js";
+
+// This allows us to test the sequential return of handlers
+// Currently `hre.created` is the only hook that is returned sequentially
+// but it returns void. `testExample` takes and returns a string to
+// ease testing.
+declare module "../../src/types/hooks.js" {
+  interface HardhatRuntimeEnvironmentHooks {
+    testExample: (context: HookContext, input: string) => Promise<string>;
+  }
+}
```

Comment:
I wouldn't do this, as adding the declaration here is equivalent to adding it to the whole codebase. Maybe some alternative to test that they are run in a sequence could be pushing to some well known or global-ish array?

Reply:
after reading more, i get why my proposed alternative doesn't work. This is trickier than I originally understood.

Maybe we can do some casting within the test instead?

Reply:
I have removed the type augmentation, and replaced it with a small amount of casting:

https://github.com/NomicFoundation/hardhat/pull/5393/commits/f9fa705a4e2fe81c724dead0764f19f0a1fc32cf

This strikes me as a fair balance between `any` usage and clarity; but I am happy to push in a different direction.